ci/run_ci.sh

```diff
@@ -28,38 +30,32 @@ diff_list=`git diff origin/master --name-only`
 # - anything in ci has been changed
 # - anything in utils has been changed
 # Else only run tests on directories that have changed
-if [[ `echo $diff_list | grep -v / | wc -l` -gt 0 || `echo $diff_list | grep ci/` || `echo $diff_list | grep utils/` ]]
-then
+if [[ `echo ${diff_list} | grep -cv /` -gt 0 || `echo ${diff_list} | grep -E "(ci|utils|image_resources)/"` ]]; then
   echo "Running full test"
-  test_list=`ls -d */ | grep -Ev "(utils|ci|ripsaw|image_resources)/"`
+  test_list=`find * -name ci_test.sh -type f -exec dirname {} \;`
 else
   echo "Running specific tests"
   echo $diff_list
-  test_list=`echo $diff_list | awk -F "/" '{print $1}' | uniq`
+  test_list=`echo "${diff_list}" | xargs dirname | uniq`
 fi
 
-echo "Running tests in the following directories: "$test_list
+echo -e "Running tests in the following directories:\n${test_list}"
 test_rc=0
 
-for dir in `echo $test_list`
-do
-  my_dir=${dir}
-  if [ -f $my_dir/ci_test.sh ]; then
-    start_time=`date`
-    figlet "CI test for "$my_dir
-    if $my_dir/ci_test.sh
-    then
-      end_time=`date`
-      duration=`date -ud@$(($(date -ud"$end_time" +%s)-$(date -ud"$start_time" +%s))) +%T`
-      echo ${dir::-1}" | PASS | "$duration >> results.markdown
-    else
-      end_time=`date`
-      duration=`date -ud@$(($(date -ud"$end_time" +%s)-$(date -ud"$start_time" +%s))) +%T`
-      echo ${dir::-1}" | FAIL | "$duration >> results.markdown
-      test_rc=1
-    fi
-    wait_clean
+for dir in ${test_list}; do
```

Comment:
so if I understand this correctly, we got rid of the -f $my_dir/ci_test.sh because the test_list was constructed to have only directories containing a ci_test.sh.   The only other change I see here is ${dir::-1} to ${dir}, which makes sense because if "dir" is an empty string or undefined then how did it get into test_list?    it's hard to see if there are other changes here because of the indentation change, but I think that's all, right?

Reply:
Yes, this is exactly what I wanted to do.